src/Sylius/Component/Promotion/Checker/PromotionEligibilityChecker.php

```diff
@@ -63,19 +63,31 @@ public function isEligible(PromotionSubjectInterface $subject, PromotionInterfac
             return false;
         }
 
+        $eligible = true;
+        $eligibleRules = false;
         if ($promotion->hasRules()) {
             foreach ($promotion->getRules() as $rule) {
                 try {
                     if (!$this->isEligibleToRule($subject, $promotion, $rule)) {
                         return false;
                     }
+
+                    $eligibleRules = true;
                 } catch (UnsupportedTypeException $exception) {
+                    if (!$eligibleRules) {
```

Comment:
What if the first rule throws an exception? This would set eligible to false prematurely. Perhaps we should do something that's more like the [AuthorizationDecisionManager](https://github.com/symfony/Security/blob/master/Core/Authorization/AccessDecisionManager.php) and rules could be more like voters.